breacharbiter.go

```diff
@@ -443,21 +379,176 @@ secondLevelCheck:
 	// channel's retribution against the cheating counter party.
 	err = b.cfg.PublishTransaction(finalTx)
 	if err != nil {
-		brarLog.Errorf("unable to broadcast "+
-			"justice tx: %v", err)
+		brarLog.Errorf("unable to broadcast justice tx: %v", err)
+
 		if err == lnwallet.ErrDoubleSpend {
-			brarLog.Infof("Attempting to transfer HTLC revocations " +
-				"to the second level")
+			// Broadcasting the transaction failed because of a
+			// conflict either in the mempool or in chain. We'll
+			// now create spend subscriptions for all HTLC outputs
+			// on the commitment transaction that could possibly
+			// have been spent, and wait for any of them to
+			// trigger.
+			brarLog.Infof("Waiting for a spend event before " +
+				"attempting to craft new justice tx.")
 			finalTx = nil
 
-			// Txn publication may fail if we're shutting down.
-			// If so, return to avoid entering an infinite loop.
+			// spend is used to wrap the index of the output that
+			// gets spent together with the spend details.
+			type spend struct {
+				index  int
+				detail *chainntnfs.SpendDetail
+			}
+
+			// We create a channel the first goroutine that gets a
+			// spend event can signal. We make it buffered in case
+			// multiple spend events come in at the same time.
+			anySpend := make(chan struct{},
+				len(breachInfo.breachedOutputs))
+
+			// The allSpends channel will be used to pass spend
+			// events from all the goroutines that detects a spend.
+			allSpends := make(chan spend,
+				len(breachInfo.breachedOutputs))
+
+			// exit will be used to signal the goroutines that they
+			// can exit.
+			exit := make(chan struct{})
+			var wg sync.WaitGroup
+
+			// We'll now launch a goroutine for each of the HTLC
```

Comment:
If there's a better alternative for effectively `select`ing on the N spend channels, I'm all ears.